tests/phpunit/tests/wp-token-map/wpTokenMap.php

```diff
@@ -0,0 +1,377 @@
+	 * @param string       $expected_token  Which token should be found at start of search document.
+	 */
+	public function test_finds_short_matches_shorter_than_group_key_length( $map, $search_document, $expected_token ) {
+		$skip_bytes = 0;
+		$text       = 'antarctica is a continent';
+		$this->assertSame(
+			'article',
+			$map->read_token( $text, 0, $skip_bytes ),
+			"Should have matched 'a' but matched '" . substr( $text, 0, $skip_bytes ) . "' instead."
+		);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[].
+	 */
+	public static function data_short_substring_matches_of_each_other() {
+		$map = WP_Token_Map::from_array(
+			array(
+				'a'       => 'article',
+				'aa'      => 'defensive weapon',
+				'ar'      => 'country code',
+				'arizona' => 'state name',
+			)
+		);
+
+		return array(
+			'single character'    => array( $map, 'antarctica is a continent', 'a' ),
+			'duplicate character' => array( $map, 'aaaaahhhh, he exclaimed', 'aa' ),
+			'different character' => array( $map, 'argentina is a country', 'ar' ),
+			'full word'           => array( $map, 'arizona was full of copper', 'arizona' ),
+		);
+	}
+
+	/**
+	 * Ensures that Token Map searches at appropriate starting offset.
+	 *
+	 * @ticket 60698.
+	 *
+	 * @dataProvider data_html5_test_dataset
+	 *
+	 * @param string $token       Token to find.
+	 * @param string $replacement Replacement string for token.
+	 */
+	public function test_reads_token_at_given_offset( $token, $replacement ) {
+		$document = "& another {$token} & then some";
+		$map      = self::get_html5_token_map();
+
+		$skip_bytes = 0;
+		$this->assertFalse(
+			$map->read_token( $document, 0, $skip_bytes ),
+			"Shouldn't have found token at start of document."
+		);
+
+		$response = $map->read_token( $document, 10, $skip_bytes );
+
+		$this->assertSame(
+			strlen( $token ),
+			$skip_bytes,
+			"Found the wrong length for token '{$token}'."
+		);
+
+		$this->assertSame(
+			$response,
+			$replacement,
+			'Found the wrong replacement value for the token.'
+		);
+	}
+
+	/**
+	 * Ensures that all given tokens exist inside a constructed Token Map.
+	 *
+	 * @ticket 60698.
+	 *
+	 * @dataProvider data_html5_test_dataset
+	 *
+	 * @param string $token       Token to find.
+	 * @param string $replacement Not used in this test.
+	 */
+	public function test_detects_all_tokens( $token, $replacement ) {
+		$map = self::get_html5_token_map();
+
+		$this->assertTrue(
+			$map->contains( $token ),
+			"Should have found '{$token}' inside the Token Map, but didn't."
+		);
+
+		$double_escaped_token = str_replace( '&', '&amp;', $token );
+		$this->assertFalse(
+			$map->contains( $double_escaped_token ),
+			"Should not have found '{$double_escaped_token}' in Token Map, but did."
+		);
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array[].
+	 */
+	public static function data_html5_test_dataset() {
+		$html5 = self::get_test_input_array( 'HTML5' );
```

Comment:
Would it be valuable to add a test that validates that `$html5` contains a certain number of tokens? Maybe it's impossible, but if for some reason the tokens wouldn't be loaded from the data file, then non tests depending on the data provider would run.

Reply:
Running the assertion while reading the test data fits perfectly 💯

I wasn't sure about such usage in the data provider, but it's just a helper so
seems to work great.